### PR TITLE
[DO NOT MERGE] Refactor distance sampling for volumes

### PR DESCRIPTION
**Pull request by biart**
Instead of calling `march` function in `PathTracer` each time the volume is entered, we treat volume scattering in the same way we treat surface scattering, using `PathTracer::trace` function.

As we need to obtain somehow a position of the next volume scattering, instead of tracing the ray to the next surface in `PathTracer` completely, we sample the distance and create volumetric shading point. Since importance distance sampling might depend on the scene lights (with backward tracing) or on camera (with forward tracing), the whole procedure of obtaining the next shading point is delegated to the visitors.

Each distance sample contains information about the phase function at the sampled point. Since now there is a BSDF wrapper around phase functions, both direct lighting integration and the bounce itself can be processed in a general way, just like normal surface `BSDF`.

It seems that integrating direct lighting over the entire ray via several distance samples does not fit well with the entire path tracing architecture and isn't applicable to non-homogeneous volumes, though it can decrease the variance. One distance sample which is used both for determining the bounce and for the direct lighting integration allows to trace more rays per second and the overall result looks not much worse that the original one.

### CODE REVIEW

**Comment by biart** [comment]
randomwalk_vs_dipole

File: src/appleseed/foundation/math/phasefunction.cpp [hunk 1]
```diff
@@ -128,4 +128,38 @@ float IsotropicPhaseFunction::sample(const Vector3f& outgoing, const Vector2f& s
     return RcpFourPi<float>();
 }
 
+
+//
+// DwivediPhaseFunction class implementation.
+//
+
+DwivediPhaseFunction::DwivediPhaseFunction(const float v)
+    : m_v(v)
```
Comment (dictoon) [inline src/appleseed/foundation/math/phasefunction.cpp]: Indent with two spaces.

File: src/appleseed/foundation/math/phasefunction.h [hunk 2]
```diff
@@ -69,14 +69,31 @@ class HenyeyPhaseFunction
...
     float evaluate(const Vector3f& outgoing, const Vector3f& incoming) const override;
     float sample(const Vector3f& outgoing, const Vector2f& s, Vector3f& incoming) const override;
 };
 
+//
+// Dwivedi phase function.
+//
+
+class DwivediPhaseFunction final
+    : public PhaseFunction
```
Comment (dictoon) [inline src/appleseed/foundation/math/phasefunction.h]: Indent with two spaces.

File: src/appleseed/foundation/math/phasefunction.h [hunk 3]
```diff
@@ -69,14 +69,31 @@ class HenyeyPhaseFunction
...
+{
+  public:
+    explicit DwivediPhaseFunction(const float v);
+    float evaluate(const Vector3f& outgoing, const Vector3f& incoming) const override;
+    float sample(const Vector3f& outgoing, const Vector2f& s, Vector3f& incoming) const override;
+
+  private:
+    const float m_v;
+};
+
```
Comment (dictoon) [inline src/appleseed/foundation/math/phasefunction.h]: Remove one blank line.

File: src/appleseed/renderer/modeling/volume/genericvolume.cpp [hunk 4]
```diff
@@ -141,6 +141,12 @@ class GenericVolume
         return sizeof(InputValues);
     }
 
+    static void channel_mis(Spectrum& values)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/volume/genericvolume.cpp]: This function modifies its input. Maybe add a verb, like `apply_channel_mis()`?

Comment (dictoon) [reply src/appleseed/renderer/modeling/volume/genericvolume.cpp]: Should probably be made private?

File: src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp [hunk 5]
```diff
@@ -33,19 +33,21 @@
...
 #include "renderer/modeling/bsdf/bsdfsample.h"
 #include "renderer/modeling/bssrdf/sss.h"
 #include "renderer/modeling/bssrdf/bssrdfsample.h"
-#include "renderer/modeling/volume/distancesample.h"
-#include "renderer/modeling/volume/volume.h"
 
 // appleseed.foundation headers.
+#include "foundation/math/phasefunction.h"
 #include "foundation/math/sampling/mappings.h"
+#include "foundation/math/rr.h"
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp]: Order alphabetically.

File: src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp [hunk 6]
```diff
@@ -33,19 +33,21 @@
...
 #include "foundation/math/scalar.h"
 #include "foundation/math/vector.h"
 #include "foundation/utility/api/specializedapiarrays.h"
 #include "foundation/utility/containers/dictionary.h"
 
+#include "foundation/math/cdf.h"
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp]: Order alphabetically.

File: src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp [hunk 7]
```diff
@@ -135,24 +139,95 @@ namespace
...
+
+            return 1.0f - 2.0f * b * (x[0] + b * (x[1] + b * (x[2] + b * x[3])));
+        }
+
+        static float compute_rcp_diffusion_length_high_albedo(const float albedo)
+        {
+            const float a = 1.0f - albedo;
+            const float b = sqrt(3.0f * a);
+
+            const float x[5] = {
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp]: Move braces to their own lines.

File: src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp [hunk 8]
```diff
@@ -162,95 +237,208 @@ namespace
...
             const RandomWalkBSSRDFInputValues* values =
                 static_cast<const RandomWalkBSSRDFInputValues*>(data);
+            const Spectrum& extinction = values->m_precomputed.m_extinction;
+            const Spectrum& albedo = values->m_precomputed.m_albedo;
+            const Spectrum& scattering = values->m_precomputed.m_scattering;
+            const Spectrum& rcp_diffusion_length = values->m_precomputed.m_rcp_diffusion_length;
+
+            // Compute the probability of classical sampling.
+            // The probability of classical sampling is high if phase function is anisotropic.
+            const float classical_sampling_prob = 0.1f;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp]: That appears to be a true constant, hence `ClassicSamplingProb`.

File: src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp [hunk 9]
```diff
@@ -162,95 +237,208 @@ namespace
...
+                &channel_cdf[0],
+                &channel_cdf[0] + channel_cdf.size(),
+                sampling_context.next2<float>());
+
+            // Sample distance (we always use classical sampling here).
+            sampling_context.split_in_place(1, 1);
+            const float distance = sample_exponential_distribution(
+                sampling_context.next2<float>(), extinction[channel]);
+
+            // Trace the initial ray till we reach the surface from inside.
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp]: _till_ -> _until_

File: src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp [hunk 10]
```diff
@@ -162,95 +237,208 @@ namespace
...
-                    sampling_context,
-                    &values->m_volume_data,
-                    distance_sample);
-
-                assert(distance_sample.m_incoming_point->is_valid());
+                if (++n_iteration > MaxIterationsCount)
+                    return false;   // path got lost inside the object
+
+                if (n_iteration > MinRRIteration && !test_rr(sampling_context, bssrdf_sample))
+                    return false;   // sample has not passed Rusian Roulette test
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp]: _Russian_

File: src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp [hunk 11]
```diff
@@ -162,95 +237,208 @@ namespace
...
+                build_cdf_and_pdf(bssrdf_sample.m_value, channel_cdf, channel_pdf);
+                const size_t channel = sample_cdf(
+                    &channel_cdf[0],
+                    &channel_cdf[0] + channel_cdf.size(),
+                    sampling_context.next2<float>());
+                if (scattering[channel] == 0.0f || bssrdf_sample.m_value[channel] == 0.0f)
+                    return false; // path got lost inside the object
+
+                // Determine if we do Dwivedi (biased) sampling or classical sampling.
+                bool is_biased = classical_sampling_prob < sampling_context.next2<float>();
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp]: Make const.

File: src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp [hunk 12]
```diff
@@ -162,95 +237,208 @@ namespace
...
+                    mis_classical,
+                    classical_sampling_prob));
+                bssrdf_sample.m_value *= transmission;
             }
 
+            clamp_in_place(bssrdf_sample.m_value, 0.0f, 2.0f);
+
+            if (n_iteration == 1)
+                bssrdf_sample.m_value *= values->m_zero_scattering_weight;
+            if (n_iteration == 2)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp]: `else if` is probably clearer.

File: src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp [hunk 13]
```diff
@@ -262,6 +450,55 @@ namespace
...
+            const bool          transmitted,
+            Spectrum&           transmission)
+        {
+            float mis_base = 0.0f;
+            for (size_t i = 0, e = Spectrum::size(); i < e; ++i)
+            {
+                const float x = -distance * extinction[i];
+                assert(FP<float>::is_finite(x));
+                transmission[i] = std::exp(x);
+                if (!transmitted) transmission[i] *= extinction[i];
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp]: Move `if` body to its own line.

File: src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp [hunk 14]
```diff
@@ -395,6 +631,54 @@ DictionaryArray RandomWalkBSSRDFFactory::get_input_metadata() const
             .insert("use", "optional")
             .insert("default", "1.0"));
 
+    metadata.push_back(
+        Dictionary()
+        .insert("name", "zero_scattering_weight")
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp]: Indent once level with respect to `Dictionary()` above.

File: src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp [hunk 15]
```diff
@@ -395,6 +631,54 @@ DictionaryArray RandomWalkBSSRDFFactory::get_input_metadata() const
...
+        .insert("label", "Zero Scattering Weight")
+        .insert("type", "numeric")
+        .insert("min",
+            Dictionary()
+            .insert("value", "0.0")
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp]: Same here.

**Comment by dictoon** [comment]
Could you also add references to relevant papers please? I haven't seen any, but maybe I missed them.

Is this one of the paper this work is based on? https://jo.dreggn.org/home/2016_dwivedi.pdf

We definitely need to merge this as soon as possible to avoid headaches caused by diverging code bases... Even if it's not optimal, complete, or even fully working.

**Comment by biart** [comment]
I will return to this soon. Sorry for keeping this PR opened for so long.

**Pull request closed without merging**